source/utils/security.py

```diff
@@ -175,21 +204,110 @@ def isObjectAboveLockScreen(obj: "NVDAObjects.NVDAObject") -> bool:
 		# that the user has switched to a secure desktop.
 		or isinstance(obj, SecureDesktopNVDAObject)
 	):
-		return True
-
-	runningAppModules = appModuleHandler.runningTable.values()
-	lockAppModule = next(filter(_isLockAppAndAlive, runningAppModules), None)
+		return False
 
+	lockAppModule = _findLockAppModule()
 	if lockAppModule is None:
-		# lockAppModule not running/registered by NVDA yet
-		log.debugWarning(
+		# lockAppModule not running/registered by NVDA yet.
+		# This is not an expected state.
+		log.error(
 			"lockAppModule not detected when Windows is locked. "
-			"Cannot detect if object is in lock app, considering object as insecure. "
+			"Cannot detect if object is below lock app, considering object as safe. "
 		)
-	elif lockAppModule is not None and obj.processID == lockAppModule.processID:
-		return True
+		return False
 
-	return False
+	from NVDAObjects.window import Window
+	if not isinstance(obj, Window):
+		log.debug(
+			"Cannot detect if object is below lock app, considering object as safe. "
+		)
+		# Must be a window instance to get the HWNDVal, other NVDAObjects do not support this.
+		return False
+
+	topLevelWindowHandle = winUser.getAncestor(obj.windowHandle, winUser.GA_ROOT)
+	return _isObjectBelowLockScreenCheckZOrder(topLevelWindowHandle, lockAppModule.processID)
+
+
+def _isObjectBelowLockScreenCheckZOrder(objWindowHandle: int, lockAppModuleProcessId: int) -> bool:
+	"""
+	This is a risky hack.
+	If the order is incorrectly detected,
+	secure information may become accessible.
+
+	If these functions fail, where possible,
+	NVDA should make NVDA objects accessible.
+	"""
+
+	def _isWindowLockApp(hwnd: winUser.HWNDVal) -> bool:
+		windowProcessId, _threadId = winUser.getWindowThreadProcessID(hwnd)
+		return windowProcessId == lockAppModuleProcessId
+
+	try:
+		return _isWindowBelowWindowMatchesCond(objWindowHandle, _isWindowLockApp)
+	except _UnexpectedWindowCountError:
+		log.debugWarning("Couldn't find lock screen")
+		return False
+
+
+class _UnexpectedWindowCountError(Exception):
+	"""
+	Raised when a window which matches the expected condition
+	is not found by _isWindowBelowWindowMatchesCond
+	"""
+	pass
+
+
+def _isWindowBelowWindowMatchesCond(
+		window: winUser.HWNDVal,
+		matchCond: Callable[[winUser.HWNDVal], bool]
+) -> bool:
+	"""
+	This is a risky hack.
+	The order may be incorrectly detected.
+
+	@returns: True if window is below a window that matches matchCond.
+	If the first window is not found, but the second window is,
+	it is assumed that the first window is below the second window.

```

Comment:
I'm not sure this comment, and therefore perhaps the logic of this function is correct anymore. Surely the default behaviour still needs to be that we assume the first window is (not) below the second? I.e. the NVDAObject's window is not found

Reply:
this comment is incorrect. I think it was a mistake of find and replacing "above"

```suggestion
	it is assumed that the first window is above the second window.
```